// @sap-cloud-sdk/generator/dist/edmx-parser/v2/edmx-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseComplexTypesV2 = parseComplexTypesV2;
exports.parseEntitySetsV2 = parseEntitySetsV2;
exports.parseEntityTypes = parseEntityTypes;
exports.parseAssociation = parseAssociation;
exports.parseAssociationSets = parseAssociationSets;
exports.parseFunctionImportsV2 = parseFunctionImportsV2;
const common_1 = require("../common");
const generator_utils_1 = require("../../generator-utils");
/**
 * @internal
 */
function parseComplexTypesV2(root) {
    return (0, common_1.parseComplexTypesBase)(root);
}
/**
 * @internal
 */
function parseEntitySetsV2(root) {
    return (0, common_1.parseEntitySetsBase)(root);
}
/**
 * @internal
 */
function parseEntityTypes(root) {
    return (0, common_1.parseEntityTypesBase)(root);
}
/**
 * @internal
 */
function parseAssociation(root) {
    return (0, common_1.getMergedPropertyWithNamespace)(root, 'Association');
}
/**
 * @internal
 */
function parseAssociationSets(root) {
    return (0, common_1.getPropertyFromEntityContainer)(root, 'AssociationSet');
}
/**
 * @internal
 */
function parseFunctionImportsV2(root) {
    return (0, common_1.getPropertyFromEntityContainer)(root, 'FunctionImport').map(f => ({
        ...f,
        Parameter: (0, generator_utils_1.forceArray)(f.Parameter)
    }));
}
