// @sap-cloud-sdk/generator/dist/generator-without-ts-morph/request-builder/imports.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.requestBuilderImportDeclarations = requestBuilderImportDeclarations;
exports.requestBuilderImports = requestBuilderImports;
const util_1 = require("@sap-cloud-sdk/util");
const imports_1 = require("../../imports");
const imports_2 = require("../imports");
/**
 * @internal
 */
function requestBuilderImportDeclarations(entity, oDataVersion, options) {
    return [
        ...(0, imports_2.externalImportDeclarations)(entity.keys),
        (0, imports_2.odataImportDeclaration)([
            ...requestBuilderImports(entity),
            'DeserializedType',
            'RequestBuilder',
            ...(0, imports_1.propertyTypeImportNames)(entity.keys)
        ].sort(), oDataVersion),
        entityImportDeclaration(entity, options),
        ...entityKeyImportDeclaration(entity.keys, options)
    ];
}
/**
 * @internal
 */
function requestBuilderImports(entity) {
    const imports = [
        'DefaultDeSerializers',
        'DeSerializers',
        'GetAllRequestBuilder'
    ];
    if (entity.creatable) {
        imports.push('CreateRequestBuilder');
    }
    if (entity.keys.length) {
        imports.push('GetByKeyRequestBuilder');
        if (entity.updatable) {
            imports.push('UpdateRequestBuilder');
        }
        if (entity.deletable) {
            imports.push('DeleteRequestBuilder');
        }
    }
    return imports;
}
/**
 * @internal
 */
function entityImportDeclaration(entity, options) {
    const generateESM = options?.generateESM;
    return {
        names: [entity.className],
        moduleIdentifier: generateESM
            ? `./${entity.className}.js`
            : `./${entity.className}`
    };
}
/**
 * @internal
 */
function entityKeyImportDeclaration(properties, options) {
    const generateESM = options?.generateESM;
    return (0, util_1.unique)(properties
        .filter(property => property.isEnum)
        .map(property => property.jsType)).map(type => ({
        names: [type],
        moduleIdentifier: generateESM ? `./${type}.js` : `./${type}`
    }));
}
