// @sap-cloud-sdk/generator/dist/edmx-parser/v4/edmx-parser.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.joinEntityTypes = joinEntityTypes;
exports.joinComplexTypes = joinComplexTypes;
exports.joinTypesWithBaseTypes = joinTypesWithBaseTypes;
exports.parseComplexTypesV4 = parseComplexTypesV4;
exports.parseEnumTypes = parseEnumTypes;
exports.parseEntityType = parseEntityType;
exports.parseEntitySetsV4 = parseEntitySetsV4;
exports.parseOperationImports = parseOperationImports;
exports.parseOperations = parseOperations;
const voca_1 = __importDefault(require("voca"));
const common_1 = require("../common");
const generator_utils_1 = require("../../generator-utils");
// eslint-disable-next-line import/no-internal-modules
const edmx_to_vdm_util_1 = require("../../edmx-to-vdm/edmx-to-vdm-util");
/**
 * @internal
 */
function joinEntityTypes(entityType, baseType) {
    // TODO: only join properties / nav properties of the respective type
    return {
        ...entityType,
        Key: {
            PropertyRef: [...entityType.Key.PropertyRef, ...baseType.Key.PropertyRef]
        },
        Property: [...entityType.Property, ...baseType.Property],
        NavigationProperty: [
            ...entityType.NavigationProperty,
            ...baseType.NavigationProperty
        ]
    };
}
/**
 * @internal
 */
function joinComplexTypes(complexType, baseType) {
    return {
        ...complexType,
        Property: [...complexType.Property, ...baseType.Property]
    };
}
/**
 * @internal
 */
function joinTypesWithBaseTypes(types, joinTypes) {
    return types.map(type => joinTypeWithBaseType(type, types, joinTypes));
}
/**
 * Recursively adds the base type data to the current type.
 * @param type - An EDMX type that can have a base type (e.g. EntityType or ComplexType).
 * @param types - All parsed EDMX types.
 * @param joinTypes - Function to ultimately join types.
 * @returns The enriched type (type + base type).
 */
function joinTypeWithBaseType(type, types, joinTypes) {
    if (type.BaseType) {
        const baseType = types.find(e => e.Name === (0, edmx_to_vdm_util_1.stripNamespace)(type.BaseType));
        if (!baseType) {
            throw new Error(`Type ${type.BaseType} not found, but defined as BaseType of Type ${type.Name}.`);
        }
        return joinTypes(type, joinTypeWithBaseType(baseType, types, joinTypes));
    }
    return type;
}
/**
 * @internal
 */
function parseComplexTypesV4(root) {
    return joinTypesWithBaseTypes((0, common_1.parseComplexTypesBase)(root), joinComplexTypes);
}
/**
 * @internal
 */
function parseEnumTypes(root) {
    return (0, common_1.getMergedPropertyWithNamespace)(root, 'EnumType').map(edmxEnumType => ({
        Name: edmxEnumType.Name,
        UnderlyingType: edmxEnumType.UnderlyingType,
        Member: (0, generator_utils_1.forceArray)(edmxEnumType.Member),
        Namespace: edmxEnumType.Namespace
    }));
}
/**
 * @internal
 */
function parseEntityType(root) {
    const entityTypes = (0, common_1.parseEntityTypesBase)(root);
    return joinTypesWithBaseTypes(entityTypes, joinEntityTypes);
}
/**
 * @internal
 */
function parseEntitySetsV4(root) {
    return (0, common_1.parseEntitySetsBase)(root).map(entitySet => ({
        ...entitySet,
        NavigationPropertyBinding: parseNavigationPropertyBinding(entitySet)
    }));
}
function parseNavigationPropertyBinding(entitySet) {
    return (0, generator_utils_1.forceArray)(entitySet.NavigationPropertyBinding);
}
/**
 * @internal
 */
function parseOperationImports(root, operationType) {
    const operations = (0, common_1.getPropertyFromEntityContainer)(root, `${voca_1.default.capitalize(operationType)}Import`);
    return operations.map(operation => ({
        ...operation,
        operationName: operation[voca_1.default.capitalize(operationType)],
        operationType
    }));
}
/**
 * @internal
 */
function parseOperations(root) {
    return [
        ...(0, common_1.getMergedPropertyWithNamespace)(root, 'Function'),
        ...(0, common_1.getMergedPropertyWithNamespace)(root, 'Action')
    ].map(operation => ({
        ...operation,
        Parameter: (0, generator_utils_1.forceArray)(operation.Parameter),
        IsBound: operation.IsBound || 'false'
    }));
}
