// @sap-cloud-sdk/generator/dist/entity/imports.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.entityImportDeclarations = entityImportDeclarations;
exports.otherEntityImports = otherEntityImports;
const ts_morph_1 = require("ts-morph");
const imports_1 = require("../imports");
const operations_1 = require("../operations");
/**
 * @internal
 */
function entityImportDeclarations(entity, service, oDataVersion, options) {
    if (oDataVersion === 'v4') {
        return (0, imports_1.mergeImportDeclarations)([
            (0, imports_1.odataImportDeclarationTsMorph)(['Entity', 'DefaultDeSerializers', 'DeSerializers', 'DeserializedType'], oDataVersion),
            ...(0, imports_1.complexTypeImportDeclarations)(entity.properties, options),
            {
                namedImports: [`${entity.className}Api`],
                moduleSpecifier: options?.generateESM
                    ? `./${entity.className}Api.js`
                    : `./${entity.className}Api`,
                kind: ts_morph_1.StructureKind.ImportDeclaration,
                isTypeOnly: true
            },
            ...(0, operations_1.operationDeclarations)(service, entity.operations),
            ...(0, imports_1.enumTypeImportDeclarations)(entity.properties, options)
        ]);
    }
    return [
        (0, imports_1.odataImportDeclarationTsMorph)(['Entity', 'DefaultDeSerializers', 'DeSerializers', 'DeserializedType'], oDataVersion),
        ...(0, imports_1.complexTypeImportDeclarations)(entity.properties, options),
        {
            namedImports: [`${entity.className}Api`],
            moduleSpecifier: options?.generateESM
                ? `./${entity.className}Api.js`
                : `./${entity.className}Api`,
            kind: ts_morph_1.StructureKind.ImportDeclaration,
            isTypeOnly: true
        },
        ...(0, imports_1.enumTypeImportDeclarations)(entity.properties, options)
    ];
}
/**
 * @internal
 */
function otherEntityImports(entity, service, options) {
    return Array.from(new Set(entity.navigationProperties.map(n => n.to)))
        .map(to => {
        const matchedEntity = service.entities.find(e => e.entitySetName === to);
        if (!matchedEntity) {
            throw Error(`Failed to find the entity from the service: ${JSON.stringify(service)} for entity ${entity}`);
        }
        return matchedEntity.className;
    })
        .filter(name => name !== entity.className)
        .map(name => otherEntityImport(name, options));
}
function otherEntityImport(name, options) {
    return {
        kind: ts_morph_1.StructureKind.ImportDeclaration,
        namedImports: [name, `${name}Type`],
        moduleSpecifier: options?.generateESM ? `./${name}.js` : `./${name}`
    };
}
