// @sap-ai-sdk/prompt-registry/dist/client/prompt-registry/prompt-templates-api.js
/*
 * Copyright (c) 2025 SAP SE or an SAP affiliate company. All rights reserved.
 *
 * This is a generated file powered by the SAP Cloud SDK for JavaScript.
 */
import { OpenApiRequestBuilder } from '@sap-ai-sdk/core';
/**
 * Representation of the 'PromptTemplatesApi'.
 * This API is part of the 'prompt-registry' service.
 */
export const PromptTemplatesApi = {
    _defaultBasePath: undefined,
    /**
     * List prompt templates
     * @param queryParameters - Object containing the following keys: scenario, name, version, retrieve, includeSpec.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    listPromptTemplates: (queryParameters) => new OpenApiRequestBuilder('get', '/lm/promptTemplates', {
        queryParameters
    }, PromptTemplatesApi._defaultBasePath),
    /**
     * Create or update a prompt template
     * @param body - Request body.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    createUpdatePromptTemplate: (body) => new OpenApiRequestBuilder('post', '/lm/promptTemplates', {
        body
    }, PromptTemplatesApi._defaultBasePath),
    /**
     * List prompt template history
     * @param scenario - Path parameter.
     * @param version - Path parameter.
     * @param name - Path parameter.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    listPromptTemplateHistory: (scenario, version, name) => new OpenApiRequestBuilder('get', '/lm/scenarios/{scenario}/promptTemplates/{name}/versions/{version}/history', {
        pathParameters: { scenario, version, name }
    }, PromptTemplatesApi._defaultBasePath),
    /**
     * Get prompt template by UUID
     * @param promptTemplateId - Path parameter.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    getPromptTemplateByUuid: (promptTemplateId) => new OpenApiRequestBuilder('get', '/lm/promptTemplates/{promptTemplateId}', {
        pathParameters: { promptTemplateId }
    }, PromptTemplatesApi._defaultBasePath),
    /**
     * Delete prompt template
     * @param promptTemplateId - Path parameter.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    deletePromptTemplate: (promptTemplateId) => new OpenApiRequestBuilder('delete', '/lm/promptTemplates/{promptTemplateId}', {
        pathParameters: { promptTemplateId }
    }, PromptTemplatesApi._defaultBasePath),
    /**
     * Import prompt template
     * @param body - Request body.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    importPromptTemplate: (body) => new OpenApiRequestBuilder('post', '/lm/promptTemplates/import', {
        body
    }, PromptTemplatesApi._defaultBasePath),
    /**
     * Export prompt template
     * @param promptTemplateId - Path parameter.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    exportPromptTemplate: (promptTemplateId) => new OpenApiRequestBuilder('get', '/lm/promptTemplates/{promptTemplateId}/export', {
        pathParameters: { promptTemplateId }
    }, PromptTemplatesApi._defaultBasePath),
    /**
     * Parse prompt template by ID
     * @param promptTemplateId - Path parameter.
     * @param body - Request body.
     * @param queryParameters - Object containing the following keys: metadata.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    parsePromptTemplateById: (promptTemplateId, body, queryParameters) => new OpenApiRequestBuilder('post', '/lm/promptTemplates/{promptTemplateId}/substitution', {
        pathParameters: { promptTemplateId },
        body,
        queryParameters
    }, PromptTemplatesApi._defaultBasePath),
    /**
     * Parse prompt template by name and version
     * @param scenario - Path parameter.
     * @param version - Path parameter.
     * @param name - Path parameter.
     * @param body - Request body.
     * @param queryParameters - Object containing the following keys: metadata.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    parsePromptTemplateByNameVersion: (scenario, version, name, body, queryParameters) => new OpenApiRequestBuilder('post', '/lm/scenarios/{scenario}/promptTemplates/{name}/versions/{version}/substitution', {
        pathParameters: { scenario, version, name },
        body,
        queryParameters
    }, PromptTemplatesApi._defaultBasePath)
};
