// @sap-cloud-sdk/odata-common/dist/request/odata-batch-request-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ODataBatchRequestConfig = void 0;
const node_crypto_1 = require("node:crypto");
const odata_request_config_1 = require("./odata-request-config");
/**
 * OData batch request configuration for an entity type.
 */
class ODataBatchRequestConfig extends odata_request_config_1.ODataRequestConfig {
    /**
     * Creates an instance of ODataBatchRequestConfig.
     * @param defaultBasePath - The default OData service path.
     * @param boundary - Request boundary for separation of subrequests. Defaults to an autogenerated value.
     */
    constructor(defaultBasePath, boundary = `batch_${(0, node_crypto_1.randomUUID)()}`) {
        super('post', defaultBasePath, {
            'content-type': `multipart/mixed; boundary=${boundary}`,
            accept: 'multipart/mixed'
        });
        this.defaultBasePath = defaultBasePath;
        this.boundary = boundary;
        this.subRequestPathType = 'relativeToService';
    }
    withSubRequestPathType(subRequestPathType) {
        this.subRequestPathType = subRequestPathType;
    }
    resourcePath() {
        return '$batch';
    }
    queryParameters() {
        return {};
    }
}
exports.ODataBatchRequestConfig = ODataBatchRequestConfig;
