// @sap-cloud-sdk/generator-common/dist/ts-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultTsConfig = void 0;
exports.formatTsConfig = formatTsConfig;
exports.readCustomTsConfig = readCustomTsConfig;
exports.tsconfigJson = tsconfigJson;
const fs_1 = require("fs");
const path_1 = require("path");
const util_1 = require("@sap-cloud-sdk/util");
const { readFile, lstat } = fs_1.promises;
/**
 * @internal
 */
const defaultTsConfig = (generateESM) => ({
    compilerOptions: {
        target: 'es2021',
        module: generateESM ? 'nodenext' : 'commonjs',
        lib: ['esnext'],
        declaration: true,
        declarationMap: false,
        sourceMap: true,
        diagnostics: true,
        moduleResolution: generateESM ? 'nodenext' : 'node',
        esModuleInterop: true,
        inlineSources: false,
        strict: true
    },
    include: ['**/*.ts'],
    exclude: ['dist/**/*', 'test/**/*', '**/*.spec.ts', 'node_modules/**/*']
});
exports.defaultTsConfig = defaultTsConfig;
/**
 * @internal
 */
function formatTsConfig(generateESM) {
    return JSON.stringify((0, exports.defaultTsConfig)(generateESM), null, 2) + util_1.unixEOL;
}
/**
 * @internal
 */
async function readCustomTsConfig(configPath) {
    try {
        if ((await lstat(configPath)).isDirectory()) {
            configPath = (0, path_1.resolve)(configPath, 'tsconfig.json');
        }
        return await readFile(configPath, 'utf8');
    }
    catch (err) {
        throw new util_1.ErrorWithCause(`Could not read tsconfig.json at ${configPath}.`, err);
    }
}
/**
 * Build a tsconfig.json file as string.
 * If transpile is true or tsconfig is provided, return the appropriate config.
 * @param transpile - Whether to transpile.
 * @param tsconfig - Path to custom tsconfig file.
 * @param generateESM - Whether to generate ES modules instead of CommonJS.
 * @returns The serialized tsconfig.json contents.
 * @internal
 */
async function tsconfigJson(transpile, tsconfig, generateESM) {
    if (transpile || tsconfig) {
        return tsconfig
            ? readCustomTsConfig(tsconfig)
            : formatTsConfig(generateESM);
    }
}
