// @sap-cloud-sdk/generator/dist/edmx-to-vdm/v4/operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.filterAndTransformOperations = filterAndTransformOperations;
exports.generateUnboundOperations = generateUnboundOperations;
exports.generateBoundOperations = generateBoundOperations;
const util_1 = require("@sap-cloud-sdk/util");
const edmx_parser_1 = require("../../edmx-parser");
const swagger_parser_1 = require("../../swagger-parser");
const common_1 = require("../common");
const edmx_to_vdm_util_1 = require("../edmx-to-vdm-util");
const operation_util_1 = require("./operation-util");
const logger = (0, util_1.createLogger)({
    package: 'generator',
    messageContext: 'operation'
});
const extractResponse = (response) => `${response}.value`;
function splitOperationsByMissingImports(operationImports, operations) {
    return operationImports.reduce(([withoutOperations, withOperations], curr) => {
        const operation = (0, operation_util_1.findOperationByImportName)(operations, curr.operationName);
        if (operation) {
            withOperations.push({
                Name: curr.Name,
                operationName: curr.operationName,
                Parameter: operation.Parameter,
                ReturnType: operation.ReturnType,
                IsBound: operation.IsBound.toLowerCase() === 'true',
                operationType: curr.operationType
            });
        }
        else {
            withoutOperations.push(curr);
        }
        return [withoutOperations, withOperations];
    }, [[], []]);
}
function extractEntitySetName(type) {
    const components = type.split('.');
    return components.pop();
}
function buildBoundOperation(currentOperation, edmxEntitySetName) {
    return {
        ...currentOperation,
        entitySetName: edmxEntitySetName,
        Parameter: currentOperation.Parameter.slice(1)
    };
}
function splitBoundOperationsByMissingParameters(operations) {
    return operations.reduce(([validOperations, operationsWithoutRequiredParameters], curr) => {
        if (!curr.Parameter.length) {
            return [
                validOperations,
                [...operationsWithoutRequiredParameters, curr]
            ];
        }
        const edmxEntitySetName = extractEntitySetName(curr.Parameter[0].Type);
        if (edmxEntitySetName) {
            return [
                [...validOperations, buildBoundOperation(curr, edmxEntitySetName)],
                operationsWithoutRequiredParameters
            ];
        }
        return [validOperations, [...operationsWithoutRequiredParameters, curr]];
    }, [[], []]);
}
/**
 * @internal
 * Joins the operation and operation Import.
 * Filters out all operations which do not have a OperationImport
 * Filters out all bound operations without a parameter and extracts the entity set name from the first parameter
 * It also removes the first parameter which contains only the entity information
 */
function filterAndTransformOperations(operationImports, operations, isBound) {
    const [withoutOperations, withOperations] = splitOperationsByMissingImports(operationImports, operations);
    const relevantOperations = withOperations.filter(operation => operation.IsBound === isBound);
    validateOperationImports(withoutOperations, withoutOperations[0]?.operationType);
    if (isBound) {
        return validateBoundOperationParameters(relevantOperations);
    }
    return relevantOperations;
}
function validateBoundOperationParameters(relevantOperations) {
    const [validOperations, operationsWithoutRequiredParameters] = splitBoundOperationsByMissingParameters(relevantOperations);
    if (operationsWithoutRequiredParameters.length) {
        logger.warn(`No parameter for bound operation which is needed to find the related entity. Skipping code generation: ${operationsWithoutRequiredParameters
            .map(operation => operation.operationName)
            .join(', \n')}`);
    }
    return validOperations;
}
function validateOperationImports(withoutOperation, operationType) {
    if (withoutOperation.length) {
        logger.warn(`Could not find ${operationType}s referenced by the following ${operationType} imports. Skipping code generation: ${withoutOperation
            .map(operationImport => `${operationImport.Name} => ${operationImport.operationName}`)
            .join(', \n')}`);
    }
}
/**
 * @internal
 */
function generateUnboundOperations(serviceMetadata, serviceName, entities, complexTypes, formatter) {
    return generateOperations(serviceMetadata, serviceName, entities, complexTypes, formatter);
}
/**
 * @internal
 */
function generateBoundOperations(serviceMetadata, serviceName, entities, complexTypes, formatter, edmxBindingEntitySetName, bindingEntityClassName) {
    return generateOperations(serviceMetadata, serviceName, entities, complexTypes, formatter, edmxBindingEntitySetName, bindingEntityClassName);
}
/**
 * @internal
 */
function generateOperations(serviceMetadata, serviceName, entities, complexTypes, formatter, edmxBindingEntitySetName, className) {
    const operations = (0, edmx_parser_1.parseOperations)(serviceMetadata.edmx.root);
    const operationImports = [
        ...(0, edmx_parser_1.parseOperationImports)(serviceMetadata.edmx.root, 'function'),
        ...(0, edmx_parser_1.parseOperationImports)(serviceMetadata.edmx.root, 'action')
    ];
    const joinedOperationData = filterAndTransformOperations(operationImports, operations, !!edmxBindingEntitySetName)
        .filter(operation => !edmxBindingEntitySetName ||
        edmxBindingEntitySetName === operation.entitySetName)
        // TODO 1571 remove when supporting entity type as parameter
        .filter(operation => !(0, edmx_to_vdm_util_1.hasUnsupportedParameterTypes)(operation));
    return joinedOperationData.map(operation => {
        const httpMethod = operation.operationType === 'function' ? 'get' : 'post';
        const swaggerDefinition = (0, swagger_parser_1.getSwaggerDefinitionForOperation)(operation.Name, httpMethod, serviceMetadata.swagger);
        return {
            ...(0, common_1.transformOperationBase)(operation, operation.Parameter, operation.operationType, swaggerDefinition, formatter, edmxBindingEntitySetName),
            httpMethod,
            returnType: (0, common_1.parseOperationReturnType)(operation.ReturnType, entities, complexTypes, extractResponse, serviceName),
            entityClassName: className
        };
    });
}
